src/Nethermind/Nethermind.OpcodeTracing.Plugin/Output/TraceOutputWriter.cs

```diff
@@ -0,0 +1,87 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Text.Json;
+using Nethermind.Logging;
+
+namespace Nethermind.OpcodeTracing.Plugin.Output;
+
+/// <summary>
+/// Writes opcode trace output to JSON files.
+/// </summary>
+public sealed class TraceOutputWriter
+{
+    private readonly ILogger _logger;
+    private readonly JsonSerializerOptions _serializerOptions = new()
+    {
+        WriteIndented = true
+    };
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="TraceOutputWriter"/> class.
+    /// </summary>
+    /// <param name="logManager">The log manager.</param>
+    public TraceOutputWriter(ILogManager logManager)
+    {
+        _logger = logManager?.GetClassLogger<TraceOutputWriter>() ?? throw new ArgumentNullException(nameof(logManager));
+    }
+
+    /// <summary>
+    /// Writes a trace output to a JSON file.
+    /// </summary>
+    /// <param name="outputDirectory">The output directory path.</param>
+    /// <param name="traceOutput">The trace output to write.</param>
+    /// <returns>The full path to the created file, or null if writing failed.</returns>
+    public async Task<string?> WriteAsync(string outputDirectory, TraceOutput traceOutput)
+    {
+        if (string.IsNullOrWhiteSpace(outputDirectory))
```

Comment:
You might want to use `PathUtils.GetApplicationResourcePath` to use relative/absolute paths better and potentially avoid issues with environments with restricted access